blog/authenticating_a_rails_api_part_1.md

```diff
@@ -0,0 +1,295 @@
+  attributes :id, :created_at, :updated_at
+
+  embed :ids
+end
+```
+
+Now let's add our `AuthenticationSerializer` and return the user's token and expiry:
+
+```ruby
+# app/serializers/authentication_serializer.rb
+
+class AuthenticationSerializer < BaseSerializer
+  attributes :email, :authentication_token, :authentication_token_expires_at
+end
+```
+
+Our happy path request spec should now be passing!
+
+#### Step 6: Handle error cases
+
+Finally, we need to make sure that error cases (such as a user providing a pre-existing email) are handled properly.
+
+In this case, we want to return a "422 - Unprocessable entity" response along with helpful error messages.  Let's add our integration test:
+
+```ruby
+# /spec/requests/v1/users_requests_spec.rb
+
+context 'with errors' do
+  context 'such as a pre-existing email' do
+    it 'returns a 422 response and JSON for errors' do
+      existing_user = create(:user)
+
+      user_attributes = {
+        user: {
+          email: existing_user.email,
+          password: user.password_digest
+        }
+      }.to_json
+
+      post(users_url, user_attributes, accept_headers)
+
+      expect(response).to have_http_status :unprocessable_entity
+      expect(json_value_at_path('errors/0')).to eq 'Email has already been taken'
+    end
+  end
+end
+```
+
+And now let's make it pass:
+
+```ruby
+# app/controllers/v1/users_controller.rb
+
+def create
+  user = SignUpUser.perform(user_params)
+
+  if user.save
+    render json: user, serializer: AuthenticationSerializer, root: :user
+  else
+    render json: { errors: user.errors.full_messages },
+           status: :unprocessable_entity
+  end
+end
+```
+
+([commit](https://github.com/IntrepidPursuits/api-authentication-demo/commit/6559f931049bd163a2e6649ffeb358d7929c02ec))
+
+We could leave it like this and be done with our feature.  But since we nearly always use this pattern in `create` actions of trying to save an object, and rendering errors if it fails, let's instead raise an error and rescue that error in our base controller:
+
+```ruby
+# app/services/sign_up_user.rb
+
+def sign_up_user
+  user.tap do |user|
+    user.reset_token!
+    user.save! # raise error if validation fails
+  end
+end
+```
+
+```ruby
+# app/controllers/v1/users_controller.rb
+
+def create
+  user = SignUpUser.perform(user_params)
+  render json: user, serializer: AuthenticationSerializer, root: :user
+end
+```
+
+```ruby
+# app/controllers/application_controller.rb
+
+class ApplicationController < ActionController::Base
+  # ...
+
+  rescue_from ActiveRecord::RecordInvalid do |exception|
+    render json: { errors: exception.message }, status: :unprocessable_entity
+  end
+end
+```
+
+[commit](https://github.com/IntrepidPursuits/api-authentication-demo/commit/fa1920508491249459183827417aba3d72f5b56e)
+
+### That's it!
+
+Congratulations, you can now sign up users.
+
+Let us know if you have feedback or questions!  And stay tuned for our next post in the series.
```

Comment:
Looks great! The only thing that strikes me is that there's quite a bit of code in here, and not a lot of discussion - there's an implicit assumption that someone reading this blog understands everything you are doing, which may not be the case. It wouldn't hurt to write a bit more about how some of the code works & what it's doing (I realize you don't want to write a new Rails book in this blog post, but even broad strokes would be good). Otherwise, we are encouraging a lot of blind copy/pasting, which is never a good idea.